internal/trie/node/branch_encode.go

```diff
@@ -0,0 +1,231 @@
+	"github.com/ChainSafe/gossamer/internal/trie/pools"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+// ScaleEncodeHash hashes the node (blake2b sum on encoded value)
+// and then SCALE encodes it. This is used to encode children
+// nodes of branches.
+func (b *Branch) ScaleEncodeHash() (encoding []byte, err error) {
+	buffer := pools.DigestBuffers.Get().(*bytes.Buffer)
+	buffer.Reset()
+	defer pools.DigestBuffers.Put(buffer)
+
+	err = b.hash(buffer)
+	if err != nil {
+		return nil, fmt.Errorf("cannot hash node: %w", err)
+	}
+
+	encoding, err = scale.Marshal(buffer.Bytes())
+	if err != nil {
+		return nil, fmt.Errorf("cannot scale encode hashed node: %w", err)
+	}
+
+	return encoding, nil
+}
+
+func (b *Branch) hash(digestBuffer io.Writer) (err error) {
+	encodingBuffer := pools.EncodingBuffers.Get().(*bytes.Buffer)
+	encodingBuffer.Reset()
+	defer pools.EncodingBuffers.Put(encodingBuffer)
+
+	err = b.Encode(encodingBuffer)
+	if err != nil {
+		return fmt.Errorf("cannot encode leaf: %w", err)
+	}
+
+	// if length of encoded leaf is less than 32 bytes, do not hash
+	if encodingBuffer.Len() < 32 {
+		_, err = digestBuffer.Write(encodingBuffer.Bytes())
+		if err != nil {
+			return fmt.Errorf("cannot write encoded branch to buffer: %w", err)
+		}
+		return nil
+	}
+
+	// otherwise, hash encoded node
+	hasher := pools.Hashers.Get().(hash.Hash)
+	hasher.Reset()
+	defer pools.Hashers.Put(hasher)
+
+	// Note: using the sync.Pool's buffer is useful here.
+	_, err = hasher.Write(encodingBuffer.Bytes())
+	if err != nil {
+		return fmt.Errorf("cannot hash encoded node: %w", err)
+	}
+
+	_, err = digestBuffer.Write(hasher.Sum(nil))
+	if err != nil {
+		return fmt.Errorf("cannot write hash sum of branch to buffer: %w", err)
+	}
+	return nil
+}
+
+// Encode encodes a branch with the encoding specified at the top of this package
+// to the buffer given.
+func (b *Branch) Encode(buffer Buffer) (err error) {
+	if !b.Dirty && b.Encoding != nil {
+		_, err = buffer.Write(b.Encoding)
+		if err != nil {
+			return fmt.Errorf("cannot write stored encoding to buffer: %w", err)
+		}
+		return nil
+	}
+
+	err = b.encodeHeader(buffer)
+	if err != nil {
+		return fmt.Errorf("cannot encode header: %w", err)
+	}
+
+	keyLE := codec.NibblesToKeyLE(b.Key)
+	_, err = buffer.Write(keyLE)
+	if err != nil {
+		return fmt.Errorf("cannot write encoded key to buffer: %w", err)
+	}
+
+	childrenBitmap := common.Uint16ToBytes(b.ChildrenBitmap())
+	_, err = buffer.Write(childrenBitmap)
+	if err != nil {
+		return fmt.Errorf("cannot write children bitmap to buffer: %w", err)
+	}
+
+	if b.Value != nil {
+		bytes, err := scale.Marshal(b.Value)
+		if err != nil {
+			return fmt.Errorf("cannot scale encode value: %w", err)
+		}
+
+		_, err = buffer.Write(bytes)
+		if err != nil {
+			return fmt.Errorf("cannot write encoded value to buffer: %w", err)
+		}
+	}
+
+	const parallel = false // TODO
```

Comment:
This is done in #2081 I updated the comment.

It was left disabled since this implementation would use potentially a lot of stack memory due to the high number / unbound number of goroutines.